zerver/decorator.py

```diff
@@ -882,13 +882,13 @@ def _wrapped_view_func(
         *args: ParamT.args,
         **kwargs: ParamT.kwargs,
     ) -> HttpResponse:
-        if not skip_rate_limiting:
-            rate_limit(request)
-
         if not request.user.is_authenticated:
             raise UnauthorizedError()
 
         user_profile = request.user
+        if not skip_rate_limiting:
+            rate_limit_user(request, user_profile, domain="api_by_user")
```

Comment:
Oh, I know, this change is breaking tests, because it makes rate limiting get applied in ordinary tests. When we were calling `rate_limit(request)` prior to this commit, it was fine because it has the 
```
    if not settings.RATE_LIMITING:
        return
```

so it's skipped in tests.

Reply:
I think we could just change this to do `rate_limit(request)` (so this commit would be a pure re-ordering), because it will choose the desired codepath given that the user is guaranteed `.is_authenticated` in this line

Reply:
I see, so essentially the problem here is that we don't want `rate_limit_user` to be called directly because we also rely on some additional checks in `rate_limit`. Do we want to rename them to `_rate_limit_*` to hint that these are implementation details?

Reply:
I think that's potentially fine. Arguably, we should inline things so that `rate_limit_user` checks the setting; we have this sort of abstraction violation in `decorator.py` already for webhook views.

```
            if settings.RATE_LIMITING:                                                                   
                rate_limit_user(request, user_profile, domain="api_by_user")                             
```

But I think skipping that for now so we can merge this makes sense.